### PR TITLE
Core/Entities: remove grid-wide visibility from setActive and implement another method for it

### PR DESCRIPTION
**Pull request by Wyreth**
**Changes proposed:**

An attempt to make setActive less troublesome by moving the grid-wide visibility (which is never used by anything on retail as far as I know, but I might be wrong on that) to another bool.

**Target branch(es):**

- [x] 3.3.5
- [x] master

**Tests performed:** works fine.

### CODE REVIEW

File: src/server/game/Entities/Object/Object.h [hunk 1]
```diff
@@ -414,6 +414,8 @@ class TC_GAME_API WorldObject : public Object, public WorldLocation
 
         bool isActiveObject() const { return m_isActive; }
         void setActive(bool isActiveObject);
+        bool isVisibleFromFarAway() const { return m_isVisibleFromFarAway; }
+        void setVisibleFromFarAway(bool isVisibleFromFarAway);
```
Comment (ccrs) [inline src/server/game/Entities/Object/Object.h]:
```
SetVisibleFar
SetFarVisible
```
keep it simple

Comment (Wyreth) [reply src/server/game/Entities/Object/Object.h]: Done, thanks!

File: src/server/game/Entities/Object/Object.cpp [hunk 2]
```diff
@@ -1058,6 +1058,14 @@ void WorldObject::setActive(bool on)
     }
 }
 
+void WorldObject::SetFarVisible(bool on)
+{
+    if (GetTypeId() == TYPEID_PLAYER)
+        return;
+
+    m_IsFarVisible = on;
```
Comment (tkrokli) [inline src/server/game/Entities/Object/Object.cpp]:
```cpp
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Entities/Object/Object.cpp:1066:5:

 fatal error: use of undeclared identifier 'm_IsFarVisible'; did you mean 'm_isFarVisible'?
    m_IsFarVisible = on;
    ^~~~~~~~~~~~~~
    m_isFarVisible

/home/travis/build/TrinityCore/TrinityCore/src/server/game/Entities/Object/Object.h:449:14:

 note: 'm_isFarVisible' declared here
        bool m_isFarVisible;
             ^
```

Comment (Wyreth) [reply src/server/game/Entities/Object/Object.cpp]: Ouch, fixed. Thanks!

**Comment by tkrokli** [comment]
BTW, the AppVeyor fail is caused by something in the latest changes in `src/server/game/Battlegrounds/Zones/BattlegroundAB.h` and nearby affected .cpp files. Something about `conversion from 'double' to 'const float' requires a narrowing conversion`.

**Comment by offl** [comment]
c2f152d903cbbf16271189b86cfebd7eacce09f5?

**Comment by Shauren** [comment]
Not enough. There might be setActive uses in scripts that used it exactly for visibility distance. I know of at least one such use - https://github.com/TrinityCore/TrinityCore/blob/618cd8cd49fd3abd35d0ddd3fb7f379b0a7941b6/src/server/scripts/Northrend/IcecrownCitadel/instance_icecrown_citadel.cpp#L990

**Comment by r00ty-tc** [comment]
Can I suggest that rather than adding extra functions, just adapt the existing setActive:

```c++
 void setActive(bool isActiveObject, bool extendVisibility = true);
```

That way, existing uses will still provide the extra visibility, but now it's also possible to only use it for the activity aspect.

Could also add a parameter to setActive on SAI to use the new mode too.

**Comment by Shauren** [comment]
No, that is not a good idea, we want it to default to false (check and fix existing uses)

**Comment by r00ty-tc** [comment]
Then, the two functions are fine and it's a case of find all instances of setActive, make a judgement call.

This is a lesson to @Wyreth. Nothing is as simple as it seems. 😆

**Comment by Shauren** [comment]
*cough* `m_respawnCompatibilityMode`

**Comment by Wyreth** [comment]
Nothing is ever as easy as we'd like it to be 😄 

I'd like to keep the two separate methods if possible - makes it easier to extend this to custom visibility values eventually™

About instances, isn't everything supposed to have max visibility range in instanced content? I vaguely recall being able to /target final bosses from the entrance in retail, although I wouldn't bet on it.

**Comment by r00ty-tc** [comment]
@Wyreth the issue is, that it's fine to make this change. But, if we're visiting this, we should find ALL places setActive is used, and attempt to derive whether it was set for activation purposes, visibility purposes, or both. Then change those locations too.

This probably should include checking SAI, since if you make this change, suddenly all places where SAI sets something active, will no longer also set the visibility too. In most cases I expect that will be fine, but there might be a few exceptions.

**Comment by Wyreth** [comment]
@r00ty-tc yeah, you're right.

Hopefully in almost every case the increased visibility is a bug and not intended - setActive is mostly used to keep the grid active so that the NPC can do its own thing without being interrupted. There's only a few exceptions I can think of (Fel Reaver, Putridus the Ancient and Wintergrasp walls, neither of which is currently affected, and the elevator @Shauren mentioned).

I've checked both SAI and core scripts - added SetFarVisible calls where I think it can be helpful to avoid issues. No SAI script needs it as far as I can tell - in all cases it's used to avoid interruptions in the script rather than for visibility.

**Comment by Aokromes** [comment]
I remember one quest on azuremyst isle or the other draenei isles, where you have to summon a drake or something like tha and that uses SAI IIRC.

**Comment by r00ty-tc** [comment]
Ah, I think it might be the SAI I did for the elite dragon group quest.. Razor something?

**Comment by Wyreth** [comment]
Yep I know [that dragon](http://www.wowhead.com/quest=9689), but it shouldn't be visible from further than normal as far as I can tell :P unless the dragon not being visible affects waypoint movement, at which I'd just go sulk in a corner or something.

**Comment by r00ty-tc** [comment]
https://github.com/TrinityCore/TrinityCore/issues/8988#issuecomment-81966255

That one there.

**Comment by r00ty-tc** [comment]
Yeah, could be right that actually visibility shouldn't be changed there.

**Comment by Wyreth** [comment]
Turns out he should actually be visible from any distance when spawned, checked other videos where it's clearly visible.

Good find, back to the drawing board :P

File: src/server/scripts/Northrend/IcecrownCitadel/instance_icecrown_citadel.cpp [hunk 3]
```diff
@@ -1002,9 +1002,15 @@ class instance_icecrown_citadel : public InstanceMapScript
                         // set the platform as active object to dramatically increase visibility range
                         // note: "active" gameobjects do not block grid unloading
                         if (GameObject* precipice = instance->GetGameObject(ArthasPrecipiceGUID))
+                        {
                             precipice->setActive(state == IN_PROGRESS);
```
Comment (Shauren) [inline src/server/scripts/Northrend/IcecrownCitadel/instance_icecrown_citadel.cpp]: active should be removed here - being the author of this script i guarantee it

Comment (Wyreth) [reply src/server/scripts/Northrend/IcecrownCitadel/instance_icecrown_citadel.cpp]: Only for precipice, or for the platform too?

**Comment by Wyreth** [comment]
Just a small note about Razormaw - it doesn't seem to work for me, even after reverting the changes in this PR. It will start working correctly if I fly to the dragon's spawn position. But otherwise he just spawns into an unactive grid and doesn't receive updates.

**Comment by Wyreth** [comment]
Is increasing grid activation distance a bad idea? Right now it's equal to the view distance, which means any event targeting a nearby grid (like Razormaw) will not happen unless the player physically visits the nearby grids.

Playing around with WorldObject::GetGridActivationRange() returning higher values turned out to allow improving certain edge-case events (Razormaw, #20728 to mention some) as well as anything else that should happen even if outside the current nearby player's view range,, while only adding a few MBs to overall ram usage (more grids get loaded).

**Comment by r00ty-tc** [comment]
The main reason to not activate creatures outside view range is to save CPU (not calling Update() on those creatures). It's the realization of "If a tree falls and there's no-one around, does it make a sound" in wow terms, it doesn't even fall..

Now, I think this is Blizz like. Back in the end of Wrath and beginning of cata, my old realm REALLY dropped in population, and I mean REALLY low. Quite often, you'd see these same artifacts. Flying on an FTP for example it'd be fairly normal to see shaman creatures and similar casting their expired buffs back onto themselves as you passed. So, I don't think we're doing things wrong here.

I do thing there are situations when creatures need activating out of range. Always active should be VERY rate (setActive), but we should really have some larger scopes available for specific use. General use, visual range is fine and is needed to keep CPU usage down.

**Comment by Wyreth** [comment]
I see, makes sense :P I was on a fairly populated realm back then, so I don't think I ever noticed that happening.

I guess it could be solved by allowing specific creatures to override grid activation range (instead of using the default map range), that is assuming creatures themselves can activate grids (I'm honestly not sure).

I think I'll step down from related issues for now and try to get a better feeling on how the grid system works - I feel like I'm just making mostly wrong assumptions right now :P

**Comment by r00ty-tc** [comment]
The main wrong assumption is that a grid is activated. Generally grids are only used as a way of reducing the number of items in any search, and as a block of objects to load/unload.

Activation happens at a much smaller level. Cell or visible range.

**Comment by malcrom** [comment]
Certain creatures need to remain active as long as the area is active.

**Comment by Faq** [comment]
@malcrom some example would be nice.

**Comment by Aokromes** [comment]
i remember some creatures with huge paths and waiting where they start / end the path the creature reachs the end.

**Comment by r00ty-tc** [comment]
Oh it definitely works (and worked on retail as far back as BC). We should handle it, as right now people might well wait forever

File: src/server/scripts/Northrend/IcecrownCitadel/instance_icecrown_citadel.cpp (hunk 3 above)
Comment (Shauren) [reply src/server/scripts/Northrend/IcecrownCitadel/instance_icecrown_citadel.cpp]: both

Comment (Wyreth) [reply src/server/scripts/Northrend/IcecrownCitadel/instance_icecrown_citadel.cpp]: Done, thanks!

**Comment by r00ty-tc** [comment]
@Wyreth is there anything else you want to do here, have you checked all uses and applied the extended visibility whrere applicable?

**Comment by Killyana** [comment]
We need 'SetFarVisible' on the DB: creature_template, creature, gameobjet_template and gameobject

**Comment by Aokromes** [comment]
maybe custom flag or whatever.

**Comment by r00ty-tc** [comment]
@Killyana I don't think that's for this change.

That's part of the visibility/activity scope overhaul that I'd like to do or someone else can (just say so else there will be duplication). There will be some way (either addon or on template/creature instance) to define these parameters (and it can be done in code).

This is just a change for the setActive for the interim for current scripting use.

**Comment by dikkedeur** [comment]
i came here since i was looking for a sollution for the quest razormaw, after using the bones at the pyre you here the sound ( kick sound? ) anyway.. the dragon is not spawned or at least i dont see it.
what would be the best way to fix this, since i saw aokromes mentioning something about visibility?
is that the visibility in the config?

**Comment by Kittnz** [comment]
FYI
https://github.com/LightsHope/server/commit/c58d027807b639cc6affc1c9163a1b936bcfa4bc
https://github.com/LightsHope/server/commit/a7f8c58c6077f1a57ad808428d06f113d30caa84

**Comment by Mihapro** [comment]
👍https://imgur.com/a/X6HGP
Tested in Halls of Origination where this is also needed and it works great. When far away (without going within "active" range), creatures normally move on paths, door open - Anraphet won't come down, but that's what far-visibility is for, not loading AI, right?  

💭 Also it is yet to be discovered if there's a flag for it - all of bosses and visual bunnies I find have either unit_flag2 0x00400000 or 0x00200000 set.

File: src/server/game/Entities/Object/Object.h [hunk 4]
```diff
@@ -414,6 +414,8 @@ class TC_GAME_API WorldObject : public Object, public WorldLocation
 
         bool isActiveObject() const { return m_isActive; }
         void setActive(bool isActiveObject);
+        bool IsFarVisible() const { return m_isFarVisible; }
+        void SetFarVisible(bool m_isFarVisible);
```
Comment (Mihapro) [inline src/server/game/Entities/Object/Object.h]: m_isFarVisible => visible ❓ 💭

**Comment by jackpoz** [comment]
thanks for the PR :)

**Comment by tkrokli** [comment]
I hope @Wyreth will be online again this year to notice the merge.

**Pull request merged**